### PR TITLE
Deprecate `deferrable: true` option of `add_foreign_key`

### PR DESCRIPTION
**Pull request by alpaca-tc**
### Motivation / Background

`deferrable: true` is deprecated in favor of `deferrable: :immediate`, and will be removed in Rails 7.2.

Because `deferrable: true` and `deferrable: :deferred` are hard to understand.
Both true and :deferred are truthy values.
This behavior is the same as the deferrable option of the add_unique_key method, added in #46192.

### Detail

- As with `add_unique_key` method added in #46192, `deferrable` now should be accept only symbols.
  - `deferrable: true` is deprecated now.
- Extend `ActiveRecord::Migration[7.0]` so that `deferrable: true` is available without deprecation message in older migration files.

### Additional information

The SQL generated will be slightly different, but unaffected.
[#46192](https://github.com/rails/rails/pull/46192#issuecomment-1437158646) confirms that `DEFERRABLE` and `DEFERRABLE INITIALLY IMMEDIATE` are equivalent.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by yahonda** [comment]
I think it needs to deprecate `deferrable: true` first If the migration parent class is `ActiveRecord::Migration[7.1]`. The deprecation message should suggest like "`deferrable: true` is deprecated and will be removed in Rails 7.2. Use `deferrable: immediate`" (English grammar needs to be fixed).

`ActiveRecord::Migration[7.0]` `deferrable: true` should work as is without deprecation warnings.  Changing the SQL statement is fine like `deferrable: true` works as `deferrable: immediate` because `ADD CONSTRAINT ... DEFERRABLE` is same as `ADD CONSTRAINT ... INITIALLY IMMEDIATE`

File: guides/source/active_record_postgresql.md
Comment (yahonda) [inline guides/source/active_record_postgresql.md:591]: This statement needs rewrite because "which has the same effect." is no longer true after removing `true` here.

**Comment by alpaca-tc** [comment]
@yahonda Fixed 😄 🙏

File: guides/source/active_record_postgresql.md
Comment (yahonda) [inline guides/source/active_record_postgresql.md:591]: I'm wondering if this statement is correct. It looks like saying that `NOT DEFERRABLE` and `DEFERRABLE INITIALLY IMMEDIATE` without `SET CONSTRAINTS ALL DEFERRED` behave the same way. 

However, when I tested the deferrable unique (not foreign key, though) constraint, these two options behaved differently.


[Example1 : No option given](https://gist.github.com/yahonda/1adcf1a776a385b8ec47c9cdf5dcf08b#example1--no-option-given) raised `ERROR:  duplicate key value violates unique constraint "example1_row_col_key"`. On the other hand [Example 5 : DEFERRABLE INITIALLY IMMEDIATE](https://gist.github.com/yahonda/1adcf1a776a385b8ec47c9cdf5dcf08b#example-5--deferrable-initially-immediate) the update successfully finished.

Comment (yahonda) [reply guides/source/active_record_postgresql.md:591]: I recall https://stackoverflow.com/questions/5300307/not-deferrable-versus-deferrable-initially-immediate explains it about unique key constraints, but I am not sure if it also applies to foreign key constraints.

Comment (alpaca-tc) [reply guides/source/active_record_postgresql.md:591]: For unique constraints, we was confirmed that `NOT DEFERRABLE` performs `each row check` and `DEFERRABLE INITIALLY IMMEDIATE` performs each `statement check`.
Foreign key constraints and unique constraint internally create Trigger, so it is possible that they are handled similarly, but I have not yet read and understand the Postgres source code.
Since I am on vacation for a week, I will continue my investigation the week after next.

However, while unique keys behave differently for `each row` and `statement`, I haven't come up with a example query where a foreign key constraint would produce the same difference in behavior as a unique key constraint.
At the moment, I guess that checking at `each row` and at `each statement` will have the same result if the foreign key constraint is violated. Hopefully I'll come up with something examples while I'm on vacation.

Comment (alpaca-tc) [reply guides/source/active_record_postgresql.md:591]: In the PostgreSQL specification, there is a description of compatibility differences with SQL-Standard.

> https://postgrespro.com/docs/postgresql/12/sql-createtable#id-1.9.3.85.9.4
> Non-Deferred Uniqueness Constraints
> When a UNIQUE or PRIMARY KEY constraint is not deferrable, PostgreSQL checks for uniqueness immediately whenever a row is inserted or modified. The SQL standard says that uniqueness should be enforced only at the end of the statement; this makes a difference when, for example, a single command updates multiple key values. To obtain standard-compliant behavior, declare the constraint as DEFERRABLE but not deferred (i.e., INITIALLY IMMEDIATE). Be aware that this can be significantly slower than immediate uniqueness checking.

- SQL-Standard checks constraints at the end of **each statement**.
- However, in PostgreSQL, when a `UNIQUE` or `PRIMARY KEY` constraint is `NOT DEFERRABLE`, it is checked with **each row** instead of **each statement**.
- When the `DEFERRABLE` option is given, check with **each statement** according to the SQL standard.

If we trust this statement, it seems that the difference in behavior between `NOT DEFERRABLE` and `DEFERRABLE` is limited to `CHECK` and `PRIMARY KEY` constraints.
Therefore, the behavior is the same between `NOT DEFERRABLE` and `DEFERRABLE` for `FOREIGN KEY` constraints.

Comment (alpaca-tc) [reply guides/source/active_record_postgresql.md:591]: Wait... it says UNIQUE and PRIMARY, not CHECK.
I'm going to recheck the specification 😵

Comment (alpaca-tc) [reply guides/source/active_record_postgresql.md:591]: @yahonda  I was reading PostgreSQL code for a while. It was very difficult for me. 🤯 

Im my current understanding is that the default behavior does not change for `FOREIGN KEY`.
Only `UNIQUE` and `PRIMARY` change the default behavior for `NOT DEFERRABLE` and `DEFERRABLE`.

Here are some 📝 

- `UNIQUE INDEX` is checked on insert/update in each Row to speed up processing
- In PostgreSQL, `UNIQUE` and `PRIMARY` constraints physically create a `UNIQUE INDEX`.
  - `UNIQUE/PRIMARY` constraints of `NOT DEFERRABLE` behave like regular `UNIQUE INDEX`.
- PostgreSQL can only create a `DEFERRABLE UNIQUE INDEX` if it is associated with a `UNIQUE` or `PRIMARY` constraint.
  - The behavior of DEFERRABLE `UNIQUE INDEX`.
    - Using INDEX and Trigger to check constraints.
    - Check each statement, not each row.
- `FOREIGN KEY` does not create `UNIQUE INDEX`, but only checks constraints with Trigger, so the behavior is the same between `NOT DEFERRABLE` and `DEFERRABLE`.

Comment (yahonda) [reply guides/source/active_record_postgresql.md:591]: Thanks for the detailed information.

**Comment by yahonda** [comment]
Would you resolve the conflicts in the activerecord/CHANGELOG.md ?

**Comment by alpaca-tc** [comment]
@yahonda done 👍

File: guides/source/active_record_postgresql.md
Comment (alpaca-tc) [reply guides/source/active_record_postgresql.md]: 🙏 😭 fixed

**Comment by yahonda** [comment]
Agreed to deprecate `:deferrable:true` as it can also take `:deferred` and `:immediate`, both of which are truth values. I think it is clearer to just allow `:deferred` and `:immediate`.

I'd like hear from other committers/core members to get some more opinion/approvals.

**Review by yahonda — APPROVED** [review]
Agreed to deprecate `:deferrable:true` as it can also take `:deferred` and `:immediate`, both of which are truth values. I think it is clearer to just allow `:deferred` and `:immediate`.

I'd like hear from other committers/core members to get some more opinion/approvals.

**Comment by ghiculescu** [comment]
This PR looks great to me, could https://edgeguides.rubyonrails.org/active_record_postgresql.html#deferrable-foreign-keys get updated at the same time?

image

**Comment by alpaca-tc** [comment]
I've rebased and fixed conflict.

**Comment by alpaca-tc** [comment]
@ghiculescu The guide has also been modified in this PR 👍  Please check Files changed.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb
Comment (yahonda) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:1135]: I should have commented this in advance, I think this requirement is PostgreSQL adapter specific.
So how about creating `add_foreign_key(from_table, to_table, **options)` method in activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb to run these warnings validations?

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb
Comment (yahonda) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:1738]: Let's move this method to activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb

File: activerecord/lib/active_record/migration/compatibility.rb
Comment (yahonda) [inline activerecord/lib/active_record/migration/compatibility.rb:141]: Let's add ` if connection.adapter_name == "PostgreSQL"` condition since this logic is PostgreSQL adapter specific one.

File: activerecord/test/cases/migration/compatibility_test.rb
Comment (yahonda) [inline activerecord/test/cases/migration/compatibility_test.rb:760]: Let's execute this test only for the PostgreSQL adapter.

**Comment by alpaca-tc** [comment]
@yahonda rebased 👍

**Review by yahonda — APPROVED** [review]

**Pull request merged**
